text/3700-ptr-tag-helpers.md

```diff
@@ -0,0 +1,128 @@
+- RFC PR: [rust-lang/rfcs#3700](https://github.com/rust-lang/rfcs/pull/3700)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add helper methods on primitive pointer types to facilitate getting and setting the tag of a pointer.
+Intended to work with programs that make use of architecture features such as AArch64
+Top-Byte Ignore (TBI), the primary use-case being writing tagging memory allocators.
+
+# Motivation
+[motivation]: #motivation
+
+Tagged pointers are pointers in which the unused top bits are set to contain some metadata - the tag.
+No 64-bit architecture today actually uses a 64-bit address space. Most operating systems only use
+the lower 48 bits, leaving higher bits unused. The remaining bits are for the most part used to
+distinguish userspace pointers (0x00) from kernelspace pointers (0xff).
+Certain architectures provide extensions, such as TBI on AArch64, that allow programs to make use of
+those unused bits to insert custom metadata into the pointer.
+
+Currently, Rust does not acknowledge TBI and related architecture extensions that enable the use of
+tagged pointers. This could potentially cause issues in cases such as working with TBI-enabled C/C++
+components over FFI, or when writing a tagging memory allocator.
+These functions are worth including in the standard library, despite their relatively niche use case
+and relative simplicity, so that there is a single known location where Miri hooks can be called to
+update the canonical address.
+This will make it easier to modify tagged pointers without breaking the Rust memory model.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This RFC adds two methods on each primitive pointer type - `ptr.tag()` and
+`ptr.with_tag(tag: u8)`.
+
+```
+assert!(ptr.tag() == 0);
+let tagged_ptr = unsafe { ptr.with_tag(63) };
+assert!(tagged_ptr.tag() == 63);
+```
+
+The primary use-case is implementing an allocator that tags pointers before returning them to the
+caller.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Within Rust's memory model, modifying the high bits offsets the pointer outside of the bounds of
+its original allocation, making any use of it Undefined Behaviour.
+
+The `with_tag()` method is only designed to be a helper for writing tagging allocators.
+Users *must* ensure that code using this method simulates a realloc from the untagged
+address to the tagged address, and that the underlying memory is only ever accessed
+using the tagged address from there onwards. Anything short of that explicitly violates the
+Rust memory model and will cause the program to break in unexpected ways.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Because the memory model we currently have is not fully compatible with memory tagging and
+tagged pointers, setting the high bits of a pointer must be done with great care in order to
```

Comment:
That's not correct. Pointer tagging works perfectly fine and is a commonly used technique in Rust. The memory model is just not compatible with *loading from a tagged pointer without first removing the tag*.

Reply:
That's why I wrote "not fully compatible" - you can make it work, just not as smoothly as one would like given what the hardware can support.
Besides, if you remove the tag, would that not make the result a different pointer? They won't compare equal after all.
Either way, I can clarify for sure if it's not self evident.